apps/docs/pages/guides/api/securing-your-api.mdx

```diff
@@ -56,10 +63,333 @@ where usename = 'authenticator';
+  END IF;
+
+  INSERT INTO private.rate_limits (ip, request_at) VALUES (req_ip, now());
+END;
+  $$;
+```
+
+Finally, configure the `public.check_request()` function to run on every serverless API request:
+
+```sql
+ALTER ROLE authenticator
+  SET pgrst.db_pre_request = 'public.check_request';
+
+NOTIFY pgrst, 'reload config';
+```
+
+To clear old entries in the `private.rate_limits` table setup a [pg_cron](/guides/database/extensions/pg_cron) job to clean them up.
+
+</TabPanel>
+
+<TabPanel id="use-additional-api-key" label="Use additional API keys">
+
+Some applications can benefit from using additional API keys managed by the application **in addition to the [Supabase API keys](/docs/guides/api/api-keys)**. This is commonly necessary in cases like:
+
+- Applications that use the serverless API without RLS policies.
+- Applications that do not use [Supabase Auth](/auth) or any other authentication system and rely on the `anon` role.
+
+<Admonition type="hint">
+
+Using the `apikey` header with the [Supabase API keys](/docs/guides/api/api-keys) is mandatory and not configurable. If you use additional API keys, you would have to distribute both the `anon` API key and your application's custom API key.
+
+</Admonition>
+
+Outline:
+
+- Your application requires the presence of the `x-app-api-key` header when the `anon` role is used to prevent abuse of your API.
+- These API keys are stored in the `private.anon_api_keys` table, and are distributed independently.
+- Each request using the `anon` role, it will be blocked with HTTP 403 if the `x-app-api-key` header is not registered in the table.
+
+Setup the table:
+
+```sql
+CREATE TABLE private.anon_api_keys (
+  id uuid primary key,
+  -- other relevant fields
+);
+```
+
+Create the `public.check_request` function:
+
+```sql
+CREATE FUNCTION public.check_request()
+  RETURNS VOID
+  LANGUAGE plpgsql
+  SECURITY DEFINER
+  AS $$
+DECLARE
+  req_app_api_key text := current_setting('request.headers', true)::json->>'x-app-api-key';
+  is_app_api_key_registered boolean;
+BEGIN
+  IF current_role <> 'anon' THEN
+    -- not `anon` role, allow the request to pass
+    RETURN;
+  END IF;
+
+  SELECT
+    true INTO is_app_api_key_registered
+  FROM private.anon_api_keys
+  WHERE
+    id = req_app_api_key
+  LIMIT 1;
+
+  IF is_app_api_key_registered IS TRUE THEN
+    -- api key is registered, allow the request to pass
+    RETURN;
+  END IF;
+
+  RAISE sqlstate 'PGRST' USING
+    message = json_build_object(
+      'message', 'No registered API key found in x-app-api-key header.')::text,
+    detail = json_build_object(
+      'status', 403)::text;
+END;
+  $$;
+```
+
+Finally, configure the `public.check_request()` function to run on every serverless API request:
+
+```sql
+ALTER ROLE authenticator
+  SET pgrst.db_pre_request = 'public.check_request';
+
+NOTIFY pgrst, 'reload config';
+```
+
+</TabPanel>
+
+<TabPanel id="hide-private-tables-functions" label="Hide private tables, views and functions">
```

Comment:
Hm, this is no good. The right way to do this is:

```sql
REVOKE ALL ON TABLE tbl FROM anon, authenticated, service_role;
```

Otherwise users will blindly follow this `pre-request` and then blame us being insecure when they forget to add an `_` to their tables.

Reply:
I'd suggest just removing this section

Reply:
Done.